File: features/step_definitions/money_steps.rb
def section_id(section)
  if section.match(/income/)
    '#add-income-section'
  else
    '#deduct-expense-section'
  end
end

When(/^I set the (income source|expense destination) to "([^"]*)"$/) do |kind, type|
  @money_type = type
  within section_id(kind) do
    select type, from: "Source"
  end
end


When(/^I set the (income|expense) date to ([\d\/]+)$/) do |kind, date|
  @money_date = Date.parse(date)
  within section_id(kind) do
    safe_date_fill_in "Date", @money_date
  end
end

When(/^I add an (income|expense) note to say "([^"]*)"$/) do |kind, note|
  @money_note = note
  within section_id(kind) do
    fill_in "Notes", with: note
  end
end

When(/^I submit the (income|expense)$/) do |kind|
  within section_id(kind) do
    click_button "Add Income"
  end
  wait_for_page_load
end

When(/^I select the (add income|deduct expense) section$/) do |section|
  find_link(section.titleize).click
  wait_for_remote_request
end

Then(/^I should see the add income section$/) do
  expect(page).to have_text("Add Income")
end

Then(/^I (could ){0,1}set the (income|expense) amount to (\d+)$/) do |_, kind, amount|
  @amount = amount
  within section_id(kind) do
    fill_in('Amount', with: amount)
  end
end

Then(/^I should see the (income|expense) added to (my own) ledger$/) do |kind, the_ledger|
  expect(find('#money-moves').text).to include(@amount)
  move_type = IncomeType.where(ledger: @owned_ledger, name: @money_type).first
  move = @owned_ledger.incomes.find_by(amount: @amount, notes: @money_note,
                                       money_move_type_id: move_type.id,
                                       date: @money_date)
  expect(move).not_to be_nil
end

When(/^I set the expense source to "([^"]*)"$/) do |arg1|
  pending # Write code here that turns the phrase above into concrete actions
end

Then(/^I should see the (income|expense) (.+) error$/) do |kind, field|
  pending # Write code here that turns the phrase above into concrete actions
end

Then(/^the (income|expense) form should reset$/) do |kind|
  container = find("#{section_id(kind)}")
  container.click # check if element is within screen and interactable
  form = container.find('form')
  %w{Amount Date Source Note}.each do |field|
    expect(form.find_field(field).value).to be_blank
  end
end

File: spec/cypress/support/step_definitions/money_steps.js
function titleize(string) {
  return string.split(/\s/)
    .map(word => `${word[0].toUpperCase()}${word.slice(1)}`)
    .join(" ");
}

function sectionId(kind) {
  return incomeOrExpense(
    kind, '#add-income-section', '#deduct-expense-section'
  );
}

function incomeOrExpense(kind, first = 'income', second = 'expense') {
  return kind.match(/income/) ? first : second;
}

function dateFieldFriendly(dateString) {
  return dateString.split('/').reverse().join('-');
}

When(/^I select the (add income|deduct expense) section$/, section => {
  cy.findLink(titleize(section)).click();
});

When(/^I set the (income source|expense type) to '([^']*)'$/, (kind, type) => {
  cy.wrap(type).as("moveType");
  const kindText = incomeOrExpense(kind, "Source", "Type");
  cy.get(sectionId(kind))
    .within(() => {
      cy.findField(kindText)
        .select(type);
    });
});

When(/^I (could ){0,1}set the (income|expense) amount to (\d+)$/, (_, kind, amount) => {
  cy.wrap(amount).as("amount");
  cy.get(sectionId(kind))
    .within(() => {
      cy.findField('Amount')
        .clear()
        .type(amount);
    });
});

When(/^I set the (income|expense) date to ([\d\/]+)$/, (kind, date) => {
  cy.wrap(date).as("date");
  cy.get(sectionId(kind))
    .within(() => {
      cy.findField("Date")
        .type(dateFieldFriendly(date));
    });
});

When(/^I add an (income|expense) note to say '([^']*)'$/, (kind, note) => {
  cy.wrap(note).as("notes");
  cy.get(sectionId(kind))
    .within(() => {
      cy.findField("Notes")
        .clear()
        .type(note);
    });
});

When(/^I submit the (income|expense)$/, kind => {
  const buttonText = incomeOrExpense(kind, "Add Income", "Deduct Expense");
  cy.findButton(buttonText)
    .click();
});

Then(
  /^I should see the (income|expense) added to (my own) ledger$/,
  function (kind, theLedger) {
    const typeClass = incomeOrExpense(kind, "IncomeType", "ExpenseType");
    cy.getMulti('@amount', '@notes', '@date', '@moveType', '@ownLedger')
      .spread((amount, notes, date, moveType, ownLedger) => {
        cy.get('#money-moves').should('contain', amount)
        cy.pondoScript('find_money_move', {
          ledger_id: ownLedger.id,
          move_type: moveType,
          type_class: typeClass,
          kind, amount, notes,  date
        }).its('notes').should('eql', notes)
      });
  }
);

Then(/^the (income|expense) form should reset$/, kind => {
  cy.get(sectionId(kind)).within(() => {
    [
      'Amount', 'Date', 'Notes',
      incomeOrExpense(kind, 'Source', 'Type')
    ].forEach(field => {
      cy.findField(field).should('have.value', '');
    });
  });
});

Then(/^I should see the (income|expense) (.+) error$/, (kind, field) => {
  cy
    .get('.field_with_errors')
    .get(sectionId(kind))
    .within(() => {
      cy.findField(titleize(field)).should('match', '.field_with_errors > *');
    });
});
